netkan/netkan/webhooks/github_mirror.py

```diff
@@ -0,0 +1,315 @@
+        "Zlib"              : 'http://www.zlib.net/zlib_license.html',
+        "Zope"              : 'http://old.zope.org/Resources/License.1',
+        "Unlicense"         : 'https://unlicense.org/UNLICENSE',
+    }
+
+    def __init__(self, filename):
+        self.filename = Path(filename)
+        self.contents = self.filename.read_text()
+        self._raw     = json.loads(self.contents)
+
+    def can_mirror(self):
+        return (
+            self._raw.get('kind', 'package') == 'package'
+            and self._raw['download_content_type'] in Ckan.mime_to_extension
+            and self.redistributable()
+        )
+
+    def authors(self):
+        auth = self._raw['author']
+        if isinstance(auth, list):
+            return auth
+        else:
+            return [auth]
+
+    def licenses(self):
+        lic = self._raw['license']
+        if isinstance(lic, list):
+            return lic
+        else:
+            return [lic]
+
+    def license_urls(self):
+        return [Ckan.license_urls[lic] for lic in self.licenses() if lic in Ckan.license_urls]
+
+    def redistributable(self):
+        for lic in self.licenses():
+            if lic in Ckan.redistributable_licenses:
+                return True
+        return False
+
+    def mirror_title(self):
+        return '{} - {}'.format(
+            self._raw['name'],
+            self._raw['version'],
+        )
+
+    def mirror_filename(self):
+        if not self._raw['download_hash']:
+            return None
+        return '{}-{}-{}.{}'.format(
+            self._raw['download_hash']['sha1'][0:8],
+            self._raw['identifier'],
+            self._raw['version'].replace(':', '-'),
+            Ckan.mime_to_extension[self._raw['download_content_type']],
+        )
+
+    def mirror_item(self):
+        return '{}-{}'.format(
+            self._raw['identifier'],
+            self._raw['version'].replace(':', '-'),
+        )
+
+    def mirror_description(self):
+        descr = self._raw['abstract'] + '<br>'
+        resources = self._raw.get('resources')
+        if resources:
+            homepage = resources.get('homepage')
+            if homepage:
+                descr += f'<br>Homepage: <a href="{homepage}">{homepage}</a>'
+            repository = resources.get('repository')
+            if repository:
+                descr += f'<br>Repository: <a href="{repository}">{repository}</a>'
+        lic = self.licenses()
+        if lic:
+            descr += '<br>License(s): {}'.format(' '.join(lic))
+        return descr
+
+    def download_content_type(self):
+        return self._raw['download_content_type']
+
+    def download_size(self):
+        return self._raw['download_size']
+
+    def download(self):
+        return self._raw['download']
+
+    def mirrored(self, ia):
+        results = ia.search_items(self.mirror_item())
+        return True if results else False
+
```

Comment:
I think we should have a similar pattern to the download counter here. A central metadata Class, extended for the Internet Archive.

We definitely need some tests for this.

Reply:
`CkanMirror` now extends `Ckan`, with the IA functionality split out into the former, so I think that request is satisfied.

How do I run the tests?